// @elastic/eui/es/components/datagrid/body/data_grid_body.styles.js
function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { css } from '@emotion/react';
import { euiScrollBarStyles, logicalSizeCSS } from '../../../global_styling';
var _ref = process.env.NODE_ENV === "production" ? {
  name: "jawla0-virtualized",
  styles: "scroll-padding:0;label:virtualized;"
} : {
  name: "jawla0-virtualized",
  styles: "scroll-padding:0;label:virtualized;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
export var euiDataGridBodyStyles = function euiDataGridBodyStyles(euiThemeContext) {
  return {
    euiDataGridBody: /*#__PURE__*/css(euiScrollBarStyles(euiThemeContext), ";;label:euiDataGridBody;"),
    virtualized: _ref,
    customRender: /*#__PURE__*/css(logicalSizeCSS('100%'), " overflow:auto;;label:customRender;")
  };
};
