// @elastic/eui/es/components/selectable/selectable.styles.js
function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { css } from '@emotion/react';
import { logicalCSS } from '../../global_styling';
export var euiSelectableStyles = {
  euiSelectable: process.env.NODE_ENV === "production" ? {
    name: "1sd0qv-euiSelectable",
    styles: "display:flex;flex-direction:column;label:euiSelectable;"
  } : {
    name: "1sd0qv-euiSelectable",
    styles: "display:flex;flex-direction:column;label:euiSelectable;",
    toString: _EMOTION_STRINGIFIED_CSS_ERROR__
  },
  fullHeight: /*#__PURE__*/css(logicalCSS('height', '100%'), ";;label:fullHeight;")
};
