// @elastic/eui/optimize/lib/components/form/radio/radio.styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.euiRadioStyles = void 0;
var _react = require("@emotion/react");
var _form = require("../form.styles");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

var euiRadioStyles = exports.euiRadioStyles = function euiRadioStyles(euiThemeContext) {
  var controlStyles = (0, _form.euiFormCustomControlStyles)(euiThemeContext);
  return {
    euiRadio: /*#__PURE__*/(0, _react.css)(controlStyles.wrapper, ";label:euiRadio;"),
    input: {
      euiRadio__circle: /*#__PURE__*/(0, _react.css)(controlStyles.input.fauxInput, " border-radius:50%;;label:euiRadio__circle;"),
      hasLabel: controlStyles.input.hasLabel,
      // Skip css`` className generation
      enabled: {
        selected: /*#__PURE__*/(0, _react.css)(controlStyles.input.enabled.selected, ";label:selected;"),
        unselected: /*#__PURE__*/(0, _react.css)(controlStyles.input.enabled.unselected, ";label:unselected;")
      },
      disabled: {
        selected: /*#__PURE__*/(0, _react.css)(controlStyles.input.disabled.selected, ";label:selected;"),
        unselected: /*#__PURE__*/(0, _react.css)(controlStyles.input.disabled.unselected, ";label:unselected;")
      },
      euiRadio__icon: /*#__PURE__*/(0, _react.css)(controlStyles.input.icon, ";label:euiRadio__icon;"),
      euiRadio__input: /*#__PURE__*/(0, _react.css)(controlStyles.input.hiddenInput, ";label:euiRadio__input;")
    },
    label: {
      euiRadio__label: /*#__PURE__*/(0, _react.css)(controlStyles.label.label, ";label:euiRadio__label;"),
      enabled: controlStyles.label.enabled,
      disabled: /*#__PURE__*/(0, _react.css)(controlStyles.label.disabled, ";label:disabled;")
    }
  };
};
