// @elastic/eui/optimize/es/components/flyout/flyout_resizable.styles.js
function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { css } from '@emotion/react';
import { logicalCSS } from '../../global_styling';
var _ref = process.env.NODE_ENV === "production" ? {
  name: "1vfos8p-right",
  styles: "justify-content:flex-start;label:right;"
} : {
  name: "1vfos8p-right",
  styles: "justify-content:flex-start;label:right;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
var _ref2 = process.env.NODE_ENV === "production" ? {
  name: "1ydooaf-left",
  styles: "justify-content:flex-end;label:left;"
} : {
  name: "1ydooaf-left",
  styles: "justify-content:flex-end;label:left;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
var _ref3 = process.env.NODE_ENV === "production" ? {
  name: "1cw850z-noOverlay",
  styles: "margin-inline:0;label:noOverlay;"
} : {
  name: "1cw850z-noOverlay",
  styles: "margin-inline:0;label:noOverlay;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
var _ref4 = process.env.NODE_ENV === "production" ? {
  name: "1gfwqnd-euiFlyoutResizableButton",
  styles: "position:absolute;label:euiFlyoutResizableButton;"
} : {
  name: "1gfwqnd-euiFlyoutResizableButton",
  styles: "position:absolute;label:euiFlyoutResizableButton;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
export var euiFlyoutResizableButtonStyles = function euiFlyoutResizableButtonStyles(_ref5) {
  var euiTheme = _ref5.euiTheme;
  return {
    euiFlyoutResizableButton: _ref4,
    overlay: {
      left: /*#__PURE__*/css(logicalCSS('right', 0), ";;label:left;"),
      right: /*#__PURE__*/css(logicalCSS('left', 0), ";;label:right;")
    },
    push: {
      left: /*#__PURE__*/css(logicalCSS('right', "-".concat(euiTheme.border.width.thick)), ";;label:left;"),
      right: /*#__PURE__*/css(logicalCSS('left', "-".concat(euiTheme.border.width.thick)), ";;label:right;")
    },
    noOverlay: {
      noOverlay: _ref3,
      left: _ref2,
      right: _ref
    }
  };
};
