// @elastic/eui/test-env/components/selectable/selectable_templates/selectable_template_sitewide.styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.euiSelectableTemplateSitewideStyles = void 0;
var _react = require("@emotion/react");
var _services = require("../../../services");
var _global_styling = require("../../../global_styling");
function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; } /*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
var visColors = (0, _services.euiPaletteColorBlind)();
var _ref = process.env.NODE_ENV === "production" ? {
  name: "q12130-euiSelectableTemplateSitewide__listItem",
  styles: "&:hover,&.euiSelectableListItem-isFocused{&:not([aria-disabled='true']){.euiSelectableListItem__text{text-decoration:none;}.euiSelectableTemplateSitewide__listItemTitle{text-decoration:underline;}}};label:euiSelectableTemplateSitewide__listItem;"
} : {
  name: "q12130-euiSelectableTemplateSitewide__listItem",
  styles: "&:hover,&.euiSelectableListItem-isFocused{&:not([aria-disabled='true']){.euiSelectableListItem__text{text-decoration:none;}.euiSelectableTemplateSitewide__listItemTitle{text-decoration:underline;}}};label:euiSelectableTemplateSitewide__listItem;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
var _ref2 = process.env.NODE_ENV === "production" ? {
  name: "rsb9r7-euiSelectableTemplateSitewide",
  styles: "display:block;label:euiSelectableTemplateSitewide;"
} : {
  name: "rsb9r7-euiSelectableTemplateSitewide",
  styles: "display:block;label:euiSelectableTemplateSitewide;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
var euiSelectableTemplateSitewideStyles = exports.euiSelectableTemplateSitewideStyles = function euiSelectableTemplateSitewideStyles(euiThemeContext) {
  var euiTheme = euiThemeContext.euiTheme;
  return {
    euiSelectableTemplateSitewide: _ref2,
    // Override EuiSelectable's default item underline
    euiSelectableTemplateSitewide__listItem: _ref,
    euiSelectableTemplateSitewide__optionMetasList: /*#__PURE__*/(0, _react.css)("display:block;", (0, _global_styling.logicalCSS)('margin-top', euiTheme.size.xs), " ", (0, _global_styling.euiFontSize)(euiThemeContext, 'xs'), " color:", euiTheme.colors.subduedText, ";;label:euiSelectableTemplateSitewide__optionMetasList;"),
    euiSelectableTemplateSitewide__optionMeta: /*#__PURE__*/(0, _react.css)("&:not(:last-of-type)::after{content:'\u2022';", (0, _global_styling.logicalCSS)('margin-horizontal', euiTheme.size.xs), " color:", euiTheme.colors.subduedText, ";};label:euiSelectableTemplateSitewide__optionMeta;"),
    metaTypes: {
      fontWeight: "\n        font-weight: ".concat(euiTheme.font.weight.medium, ";\n      "),
      application: /*#__PURE__*/(0, _react.css)("color:", (0, _services.makeHighContrastColor)(visColors[1])(euiTheme), ";;label:application;"),
      deployment: /*#__PURE__*/(0, _react.css)("color:", (0, _services.makeHighContrastColor)(visColors[0])(euiTheme), ";;label:deployment;"),
      article: /*#__PURE__*/(0, _react.css)("color:", (0, _services.makeHighContrastColor)(visColors[3])(euiTheme), ";;label:article;"),
      case: /*#__PURE__*/(0, _react.css)("color:", (0, _services.makeHighContrastColor)(visColors[9])(euiTheme), ";;label:case;"),
      platform: /*#__PURE__*/(0, _react.css)("color:", (0, _services.makeHighContrastColor)(visColors[5])(euiTheme), ";;label:platform;")
    }
  };
};
