// @elastic/eui/test-env/services/copy/tabular_copy.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tabularCopyMarkers = exports.OverrideCopiedTabularContent = void 0;
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

// Don't render these characters in Jest snapshots. I don't want to deal with the Kibana tests 🫠
var tabularCopyMarkers = exports.tabularCopyMarkers = {
  hiddenTab: (0, _react2.jsx)(_react.default.Fragment, null),
  hiddenNewline: (0, _react2.jsx)(_react.default.Fragment, null),
  hiddenWrapperBoundary: (0, _react2.jsx)(_react.default.Fragment, null),
  hiddenNoCopyBoundary: (0, _react2.jsx)(_react.default.Fragment, null)
};

// Don't bother initializing in Kibana Jest either
var OverrideCopiedTabularContent = exports.OverrideCopiedTabularContent = function OverrideCopiedTabularContent(_ref) {
  var children = _ref.children;
  return children;
};
