// @elastic/eui/optimize/es/components/markdown_editor/plugins/markdown_default_plugins/processing_plugins.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { Fragment, createElement } from 'react';
// Importing seemingly unused types from `unified` because the definitions
// are exported for two versions of TypeScript (3.4, 4.0) and implicit
// imports during eui.d.ts generation default to the incorrect version (3.4).
// Explicit imports here resolve the version mismatch.

import all from 'mdast-util-to-hast/lib/all';
import rehype2react from 'rehype-react';
import remark2rehype from 'remark-rehype';
import { EuiLink } from '../../../link';
import { EuiCodeBlock, EuiCode } from '../../../code';
import { EuiHorizontalRule } from '../../../horizontal_rule';
import { FENCED_CLASS } from '../remark/remark_prismjs';
import * as MarkdownTooltip from '../markdown_tooltip';
import * as MarkdownCheckbox from '../markdown_checkbox';
import { jsx as ___EmotionJSX } from "@emotion/react";
var unknownHandler = function unknownHandler(h, node) {
  return h(node, node.type, node, all(h, node));
};
var DEFAULT_COMPONENT_RENDERERS = {
  checkbox: MarkdownCheckbox.renderer,
  tooltip: MarkdownTooltip.renderer
};
export var getDefaultEuiMarkdownProcessingPlugins = function getDefaultEuiMarkdownProcessingPlugins() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
    exclude = _ref.exclude,
    linkProps = _ref.linkProps;
  var componentPluginsWithExclusions = {};
  Object.entries(DEFAULT_COMPONENT_RENDERERS).forEach(function (_ref2) {
    var _ref3 = _slicedToArray(_ref2, 2),
      excludeName = _ref3[0],
      renderer = _ref3[1];
    if (!(exclude !== null && exclude !== void 0 && exclude.includes(excludeName))) {
      var pluginName = "".concat(excludeName, "Plugin");
      componentPluginsWithExclusions[pluginName] = renderer;
    }
  });
  var plugins = [[remark2rehype, {
    allowDangerousHtml: true,
    unknownHandler: unknownHandler,
    handlers: {} // intentionally empty, allows plugins to extend if they need to
  }], [rehype2react, {
    createElement: createElement,
    Fragment: Fragment,
    components: _objectSpread({
      a: function a(props) {
        return ___EmotionJSX(EuiLink, _extends({}, props, linkProps));
      },
      code: function code(props) {
        return (
          // If there are linebreaks use codeblock, otherwise code
          /\r|\n/.exec(props.children) || props.className && props.className.indexOf(FENCED_CLASS) > -1 ? ___EmotionJSX(EuiCodeBlock, _extends({
            fontSize: "m",
            paddingSize: "s",
            isCopyable: true
          }, props)) : ___EmotionJSX(EuiCode, props)
        );
      },
      // When we use block code "fences" the code tag is replaced by the `EuiCodeBlock`.
      // But there's a `pre` tag wrapping all the `EuiCodeBlock`.
      // We want to replace this `pre` tag with a `div` because the `EuiCodeBlock` has its own children `pre` tag.
      pre: function pre(props) {
        return ___EmotionJSX("div", _extends({}, props, {
          className: "euiMarkdownFormat__codeblockWrapper"
        }));
      },
      blockquote: function blockquote(props) {
        return ___EmotionJSX("blockquote", _extends({}, props, {
          className: "euiMarkdownFormat__blockquote"
        }));
      },
      table: function table(props) {
        return ___EmotionJSX("table", _extends({
          className: "euiMarkdownFormat__table"
        }, props));
      },
      hr: function hr(props) {
        return ___EmotionJSX(EuiHorizontalRule, props);
      }
    }, componentPluginsWithExclusions)
  }]];
  return plugins;
};
export var defaultProcessingPlugins = getDefaultEuiMarkdownProcessingPlugins();
