// @elastic/eui/es/components/collapsible_nav_beta/collapsible_nav_button/collapsible_nav_button.styles.js
function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { css } from '@emotion/react';
import { logicalCSS, logicalSizeCSS } from '../../../global_styling';
import { euiHeaderVariables } from '../../header/header.styles';
var _ref = process.env.NODE_ENV === "production" ? {
  name: "msjxar-euiCollapsibleNavButton",
  styles: "&.euiButtonIcon:hover{transform:none;};label:euiCollapsibleNavButton;"
} : {
  name: "msjxar-euiCollapsibleNavButton",
  styles: "&.euiButtonIcon:hover{transform:none;};label:euiCollapsibleNavButton;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
export var euiCollapsibleNavButtonWrapperStyles = function euiCollapsibleNavButtonWrapperStyles(euiThemeContext) {
  var euiTheme = euiThemeContext.euiTheme;
  var _euiHeaderVariables = euiHeaderVariables(euiThemeContext),
    height = _euiHeaderVariables.height,
    padding = _euiHeaderVariables.padding;
  return {
    euiCollapsibleNavButtonWrapper: /*#__PURE__*/css("display:flex;align-items:center;justify-content:center;", logicalSizeCSS(height), ";;label:euiCollapsibleNavButtonWrapper;"),
    euiCollapsibleNavButton: _ref,
    left: /*#__PURE__*/css(logicalCSS('border-right', euiTheme.border.thin), " ", logicalCSS('margin-left', "-".concat(padding)), " ", logicalCSS('margin-right', padding), ";;label:left;"),
    right: /*#__PURE__*/css(logicalCSS('border-left', euiTheme.border.thin), " ", logicalCSS('margin-right', "-".concat(padding)), " ", logicalCSS('margin-left', padding), ";;label:right;")
  };
};
